### ISSUE TITLE
Support for Synchronous Replication 

### ISSUE THREAD

**Issue by michail-nikolaev** [zalando/patroni#90]
Using synchronous_standby_names (one sync replica, 0+ async replicas)

http://www.postgresql.org/docs/9.4/static/warm-standby.html#SYNCHRONOUS-REPLICATION

**Comment by michail-nikolaev** [comment]
Oops, missed it

**Comment by michail-nikolaev** [comment]
Will Patroni select exact current sync replica for new master?

**Comment by alexeyklyukin** [comment]
Right now Patroni does not support synchronous replication: particularly, what needs to be implemented is setting the application_name/syncrhonous_standby_name and making sure the we go into the synchronous mode only after we bootstrap at least one replica. As for selecting the sync replica by default as a target for the failover, it is not necessary: Patroni will select the replica that is the closest to the master by the position in the WAL recently replayed.

**Comment by katoquro** [comment]
Hello
Do you have any estimate when support of sync replication can be done? 

If I right understood how patroni works it generates and provides it own config file for postgres (like ${DATA_DIR}/postgresql.conf). So if I make any changes in postgresql.conf it will not affect cluster configuration. If it so where can I see the result configuration?

Thanks

**Comment by alexeyklyukin** [comment]
Hi,

With the changes merged today (https://github.com/zalando/patroni/pull/94) it is possible to run the synchronous replication by adding the following lines to the postgresql section of your patroni configuration:

    synchronous_commit: "on"
    synchronous_standby_names: "*"

**Comment by michail-nikolaev** [comment]
@alexeyklyukin :+1:

**Comment by drnic** [comment]
Awesome!

**Comment by drnic** [comment]
If a cluster has a sync replica and several async replicas will Patroni always failover to the sync replica? Or is that still future work to manage the two classes of replication differently?

**Comment by michail-nikolaev** [comment]
@drnic 
"Patroni will select the replica that is the closest to the master by the position in the WAL recently replayed."

So, probably sync replica, yes.

**Comment by drnic** [comment]
Lovely. Thanks for finding that.
